File: server/python/hosted_checkout/hosted_checkout_mapper.py
from utils import to_amount
from configuration.config_reader import read_config
from onlinepayments.sdk.domain.create_hosted_checkout_request import CreateHostedCheckoutRequest # type: ignore

def to_empty_dto():
    empty_dto = {
        "redirectUrl": read_config()['hostedCheckout']['redirectUrl']
    }

    return empty_dto

def to_hosted_checkout_request(hosted_checkout_request_dto):

    hosted_checkout_request_dict = {
        "order": {
            "amountOfMoney": {
                "amount": to_amount(hosted_checkout_request_dto['amount']),
                "currencyCode": hosted_checkout_request_dto['currency']
            }
        },
        "hostedCheckoutSpecificInput": {
            "returnUrl": hosted_checkout_request_dto['redirectUrl']
        }
    }

    hosted_checkout_request = CreateHostedCheckoutRequest()
    hosted_checkout_request.from_dictionary(hosted_checkout_request_dict)

    return hosted_checkout_request

File: server/ruby/app/hosted_checkout/hosted_checkout_mapper.rb
require_relative '../services/payment_utils'

require 'onlinepayments/sdk'
include OnlinePayments::SDK::Domain

class HostedCheckoutMapper
  def self.to_empty_dto
    CreateHostedCheckoutBasicDto.new(nil, nil, Rails.configuration.hostedCheckout.redirectUrl)
  end

  def self.to_create_hosted_checkout_request(create_hosted_checkout_basic_dto)
    create_hosted_checkout_request = CreateHostedCheckoutRequest.new

    order = Order.new

    amount_of_money = AmountOfMoney.new
    amount_of_money.amount = PaymentUtils.to_amount(create_hosted_checkout_basic_dto.amount)
    amount_of_money.currency_code = create_hosted_checkout_basic_dto.currency

    order.amount_of_money = amount_of_money

    create_hosted_checkout_request.order = order

    hosted_checkout_specific_input = HostedCheckoutSpecificInput.new
    hosted_checkout_specific_input.return_url = create_hosted_checkout_basic_dto.redirect_url

    create_hosted_checkout_request.hosted_checkout_specific_input = hosted_checkout_specific_input

    return create_hosted_checkout_request

  end

  def to_amount(amount)
    (amount * BigDecimal.new('100')).round(0, BigDecimal::ROUND_HALF_UP).to_i
  end
end
